channel.go

```diff
@@ -0,0 +1,139 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hashicorp/yamux"
+	"golang.org/x/sys/unix"
+)
+
+type channelType string
+
+const (
+	serialChannelName = "agent.channel.0"
+	virtIOPath        = "/sys/class/virtio-ports"
+	devRootPath       = "/dev"
+	vSockDevPath      = "/dev/vsock"
+)
+
+type channel interface {
+	getType() channelType
+}
```

Comment:
So I think we should have a different channel interface:

```Go
type channel interface {
        open() error
        close() error
        listen() (*net.Listener, error)
}
```

With that interface:

* I would implement a `newChannel() (*channel, error)` function that would implement the entire logic of checking if we are going to do vsock or serial, and respectively return the vsock or serial implementation.
* The `func (p *pod) initChannel() error` would become something like:
```Go
func (p *pod) initChannel() error {
    c, err := newChannel()
    if err != nil {
         return err
    }

    p.channel = c
    return p.channel.open()

}
``` 
* The serial `open()` implementation would create the yamux server on its serial path and store the yamux session.
* The serial `listen()` implementation would simply return the yamux session
* `func (p *pod) startGRPC() (err error)` would be simplified to:
```Go
func (p *pod) startGRPC() error {
        l, err := p.channel.listen()
       if err != nil {
          return err
       }

	grpcImpl := &agentGRPC{
		pod: p,
	}

	grpcServer := grpc.NewServer()
	pb.RegisterHyperstartServiceServer(grpcServer, grpcImpl)

	p.wg.Add(1)
	go func() {
		defer p.wg.Done()

		grpcServer.Serve(l)
	}()

	return nil
}
```

And so on...
I think the code would be simplified and more readable overall.